File: data_structures/linked_list/linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def __str__(self):
        values = []
        current = self.head
        while current:
            values.append(str(current.value))
            current = current.next
        return ' -> '.join(values)
        
    def add(self, value):
        new_node = Node(value)
        if self.head == None:
            self.head = new_node
            self.tail = new_node
            self.length = 1
        else:
            self.tail.next = new_node
            self.tail = new_node
            self.length += 1
            
    def prepend(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        if self.length == 0:
            self.tail = new_node
        self.length += 1
    
    def insert(self, index, value):
        new_node = Node(value)
        if index >= self.length:
            self.add(value)
            return
        if index == 0:
            self.prepend(value)
            return
        leading = self.traverseToIndex(index-1)
        holding = leading.next
        leading.next = new_node
        new_node.next = holding
        self.length += 1
        
    def remove(self, index):
        if index < 0 or index >= self.length:
            raise IndexError("Index out of bounds")
        elif index == 0:
            self.head = self.head.next
            if self.length == 1:
                self.tail = None
        else:
            leading = self.traverseToIndex(index-1)
            unwant = leading.next
            leading.next = unwant.next
            if index == self.length - 1:
                self.tail = leading
        self.length -= 1
        
    def reverse(self):
        prev = None
        currentNode = self.head
        self.tail = currentNode
        while currentNode:
            n = currentNode.next
            currentNode.next = prev
            prev = currentNode
            currentNode = n
        self.head = prev
        
    def traverseToIndex(self, index):
        count = 0
        currentNode = self.head
        while count != index:
            currentNode = currentNode.next
            count += 1
        return currentNode
        
    def printlist(self):
        myarr = []
        currentNode = self.head
        while currentNode:
            myarr.append(currentNode.value)
            currentNode = currentNode.next
        return myarr
        
mylist = LinkedList()
mylist.prepend(888)
mylist.add(1)
mylist.add(2)
mylist.add(3)
mylist.prepend(99)
mylist.insert(99, 88)
mylist.insert(0, 88)
mylist.insert(4, 88)
print(mylist.printlist())
mylist.remove(88)
print(mylist.printlist())
mylist.remove(0)
print(mylist.printlist())
mylist.remove(3)
print(mylist.printlist())
mylist.reverse()
print(mylist.printlist())

File: data_structures/linked_list/LinkedList.java


class Node {
    int value;
    Node next;

    Node(int value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    private Node head;
    private Node tail;
    private int length;

    LinkedList() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        Node current = head;
        while (current != null) {
            result.append(current.value).append(" -> ");
            current = current.next;
        }
        return result.length() > 0 ? result.substring(0, result.length() - 4) : "";
    }

    public void add(int value) {
        Node newNode = new Node(value);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
        length++;
    }

    public void prepend(int value) {
        Node newNode = new Node(value);
        if (length == 0) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head = newNode;
        }
        length++;
    }

    public void insert(int index, int value) {
        if (index >= length) {
            add(value);
            return;
        }
        if (index == 0) {
            prepend(value);
            return;
        }
        Node newNode = new Node(value);
        Node leading = traverseToIndex(index - 1);
        Node holding = leading.next;
        leading.next = newNode;
        newNode.next = holding;
        length++;
    }

    public void remove(int index) {
        if (index < 0 || index >= length) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }
        if (index == 0) {
            head = head.next;
            if (length == 1) {
                tail = null;
            }
        } else {
            Node leading = traverseToIndex(index - 1);
            Node unwanted = leading.next;
            leading.next = unwanted.next;
            if (index == length - 1) {
                tail = leading;
            }
        }
        length--;
    }

    public void reverse() {
        Node prev = null;
        Node currentNode = head;
        tail = head;

        while (currentNode != null) {
            Node next = currentNode.next;
            currentNode.next = prev;
            prev = currentNode;
            currentNode = next;
        }
        head = prev;
    }

    private Node traverseToIndex(int index) {
        int count = 0;
        Node currentNode = head;
        while (count != index) {
            currentNode = currentNode.next;
            count++;
        }
        return currentNode;
    }

    public int[] printList() {
        int[] result = new int[length];
        int index = 0;
        Node currentNode = head;
        while (currentNode != null) {
            result[index++] = currentNode.value;
            currentNode = currentNode.next;
        }
        return result;
    }

    public static void main(String[] args) {
        LinkedList myList = new LinkedList();
        myList.prepend(888);
        myList.add(1);
        myList.add(2);
        myList.add(3);
        myList.prepend(99);
        myList.insert(99, 88);
        myList.insert(0, 88);
        myList.insert(4, 88);

        System.out.println(java.util.Arrays.toString(myList.printList()));
        
        // Removing the element at index 88 will throw an error
        // Uncomment to see the exception:
        // myList.remove(88); 

        myList.remove(0);
        System.out.println(java.util.Arrays.toString(myList.printList()));

        myList.remove(3);
        System.out.println(java.util.Arrays.toString(myList.printList()));

        myList.reverse();
        System.out.println(java.util.Arrays.toString(myList.printList()));
        System.out.println("toString: " + myList.toString());
    }
}
